### PR TITLE
Add common used score_mod functions for templated attention

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #ISSUE_NUMBER

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/nn/attention/_score_mod.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+"""This module defines the common used score_mod functions for templated attention in PyTorch."""
+import torch
+
+
+def identity(
```
Comment (drisspg) [inline torch/nn/attention/_score_mod.py:5]: I think add these to the 'templated_attention' file

File: torch/nn/attention/_score_mod.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def alibi_bias(
+    score: torch.Tensor,
+    batch: torch.Tensor,
+    head: torch.Tensor,
+    token_q: torch.Tensor,
+    token_kv: torch.Tensor,
+) -> torch.Tensor:
+    return score + (token_q - token_kv) * head
```
Comment (drisspg) [inline torch/nn/attention/_score_mod.py:52]: Also this isn't really alib attention yet, https://github.com/drisspg/transformer_nuggets/blob/c11220bbdd14a3f139013b91beec1f8e969b506a/transformer_nuggets/flash/masks.py#L67

this would be actual alibi bias if the num_heads was captured, so not sure if we want to add these just yet

File: torch/nn/attention/_score_mod.py (hunk 1 above)
Comment (drisspg) [reply torch/nn/attention/_score_mod.py:5]: also maybe keep em underscored for now so you dont have to worry about docs

Comment (yanboliang) [reply torch/nn/attention/_score_mod.py:5]: SG!

File: torch/nn/attention/_score_mod.py (hunk 2 above)
Comment (yanboliang) [reply torch/nn/attention/_score_mod.py:52]: Yes, but this function's name is ```_alibi_bias```, is that right?

File: torch/nn/attention/_templated_attention.py [hunk 3]
```diff
@@ -90,3 +90,68 @@ def score_mod(
...
+    score: torch.Tensor,
+    batch: torch.Tensor,
+    head: torch.Tensor,
+    token_q: torch.Tensor,
+    token_kv: torch.Tensor,
+) -> torch.Tensor:
+    return score + (token_q - token_kv) * head
+
+
+def _alibi_causal(
```
Comment (Chillee) [inline torch/nn/attention/_templated_attention.py:148]: I wouldn't bother with this one - just make it from composing alibi + causal together.

File: torch/nn/attention/_templated_attention.py [hunk 4]
```diff
@@ -90,3 +90,68 @@ def score_mod(
 
     # Drop the logsumexp value since this is only needed for backwards
     return out
+
+
+"""The common used score_mod functions for templated attention in PyTorch."""
```
Comment (Chillee) [inline torch/nn/attention/_templated_attention.py:95]: Just add "some" commonly used score_mod functions.

File: test/inductor/test_templated_attention.py [hunk 5]
```diff
@@ -80,58 +94,8 @@ def run_test(self, score_mod: Callable, dtype: torch.dtype = torch.float16):
...
-
-        self.run_test(score_mod, dtype)
-
-    @supported_platform
-    @common_utils.parametrize("dtype", test_dtypes)
-    def test_alibi_causal(self, dtype: torch.dtype):
-        def score_mod(score, b, h, m, n):
-            return torch.where(m <= n, score + (m - n) * h, float("-inf"))
-
+    @common_utils.parametrize("score_mod", test_score_modes)
```
Comment (Chillee) [inline test/inductor/test_templated_attention.py:97]: `test_score_mods`

File: torch/nn/attention/_templated_attention.py [hunk 6]
```diff
@@ -90,3 +90,68 @@ def score_mod(
...
+    score: torch.Tensor,
+    batch: torch.Tensor,
+    head: torch.Tensor,
+    token_q: torch.Tensor,
+    token_kv: torch.Tensor,
+) -> torch.Tensor:
+    return torch.where(token_q <= token_kv, score + (token_q - token_kv), float("-inf"))
+
+
+def _alibi_bias(
```
Comment (Chillee) [inline torch/nn/attention/_templated_attention.py:138]: I agree with Driss this isn't correct. I think to properly support this we need https://github.com/pytorch/pytorch/pull/124444, and then we can capture in a tensor.

**Review by Chillee — APPROVED** [review]
